cmd/shim/main.go

```diff
@@ -0,0 +1,235 @@
+	for i, arg := range os.Args {
+		if arg == "--bundle" {
+			if i+1 >= len(os.Args) {
+				fmt.Printf("Missing bundle path\n")
+				return 1
+			}
+			bundleDir = os.Args[i+1]
+			break
+		}
+	}
+	if bundleDir == "" {
+		fmt.Printf("Missing bundle path\n")
+		return 1
+	}
+
+	configPath := filepath.Join(bundleDir, "config.json")
+	configBytes, err := os.ReadFile(configPath)
+	if err != nil {
+		fmt.Printf("Error reading config.json: %v\n", err)
+		return 1
+	}
+
+	var spec specs.Spec
+	if err := json.Unmarshal(configBytes, &spec); err != nil {
+		fmt.Printf("Error parsing config.json: %v\n", err)
+		return 1
+	}
+
+	// Check to see if this is a dagger exec, currently by using
+	// the presence of the dagger meta mount. If it is, set up the
+	// shim to be invoked as the init process. Otherwise, just
+	// pass through as is
+	var isDaggerExec bool
+	for _, mnt := range spec.Mounts {
+		if mnt.Destination == metaMountPath {
+			isDaggerExec = true
+			break
+		}
+	}
+	if isDaggerExec {
+		selfPath, err := os.Executable()
+		if err != nil {
+			fmt.Printf("Error getting self path: %v\n", err)
+			return 1
+		}
+		selfPath, err = filepath.EvalSymlinks(selfPath)
+		if err != nil {
+			fmt.Printf("Error getting self path: %v\n", err)
+			return 1
+		}
+		spec.Mounts = append(spec.Mounts, specs.Mount{
+			Destination: shimPath,
+			Type:        "bind",
+			Source:      selfPath,
+			Options:     []string{"rbind", "ro"},
+		})
+
+		spec.Process.Args = append([]string{shimPath}, spec.Process.Args...)
+
+		configBytes, err = json.Marshal(spec)
+		if err != nil {
+			fmt.Printf("Error marshaling config.json: %v\n", err)
+			return 1
+		}
+		if err := os.WriteFile(configPath, configBytes, 0600); err != nil {
+			fmt.Printf("Error writing config.json: %v\n", err)
+			return 1
+		}
+	}
+
+	args := []string{runcPath}
+	args = append(args, os.Args[1:]...)
+	if err := unix.Exec(runcPath, args, os.Environ()); err != nil {
+		fmt.Printf("Error execing runc: %v\n", err)
+		return 1
+	}
+	panic("congratulations: you've reached unreachable code, please report a bug!")
+}
+
+func proxyAPI(daggerHost string) (string, error) {
+	u, err := url.Parse(daggerHost)
+	if err != nil {
+		return "", err
+	}
+	proxy := httputil.NewSingleHostReverseProxy(&url.URL{
+		Scheme: "http",
+		Host:   "localhost",
+	})
+	proxy.Transport = &http.Transport{
+		DialContext: func(_ context.Context, _, _ string) (net.Conn, error) {
+			return net.Dial("unix", u.Path)
+		},
+	}
+
+	l, err := net.Listen("tcp", "localhost:0")
+	if err != nil {
+		return "", err
+	}
+	port := l.Addr().(*net.TCPAddr).Port
+
+	srv := &http.Server{
+		Handler:           proxy,
+		ReadHeaderTimeout: 10 * time.Second,
+	}
+	go srv.Serve(l)
+	return fmt.Sprintf("http://localhost:%d", port), nil
+}
+
+func internalEnv(name string) (string, bool) {
+	val, found := os.LookupEnv(name)
+	if !found {
+		return "", false
+	}
+
+	os.Unsetenv(name)
+
+	return val, true
+}
+
+func main() {
+	if os.Args[0] == shimPath {
```

Comment:
Can you add a comment explaining this sorcery?

If I understand correctly, we're getting called twice: once as `dagger-shim` which prepares calling ourselves again, this time as `_shim`, which triggers a different code path

Also, instead of documenting, WDYT if we made it more explicit by having `cmd/shim` and `cmd/oci-worker`? Since we're injecting a binary, we might as well inject 2.

Reply:
Sounds good to me!

Reply:
Updated with a lot more comments, should be more clear now hopefully